File: frontend/src/types/sop.ts
export type SopProcessingStatus =
  | "queued"
  | "formatting"
  | "indexing"
  | "ready"
  | "failed";

export interface SopCategory {
  id: string;
  slug: string;
  name: string;
  description?: string | null;
}

export interface SopProcessStep {
  id?: string;
  type?: string;
  instruction?: string;
  tool?: string;
  inputs?: Record<string, unknown>;
}

export interface SopProcess {
  id: string;
  process_key: string;
  title: string;
  description: string;
  trigger_phrases: string[];
  steps: SopProcessStep[];
  tools: string[];
  sort_order: number;
}

export interface SopDocumentListItem {
  id: string;
  title: string;
  category_id: string;
  category_slug?: string | null;
  category_name?: string | null;
  summary: string | null;
  enabled: boolean;
  processing_status: SopProcessingStatus;
  processing_error: string | null;
  process_count: number;
  tool_warnings?: string[] | null;
  updated_at: string;
}

export interface SopDocument extends SopDocumentListItem {
  raw_text: string;
  markdown: string | null;
  content_hash: string;
  embedding_model: string | null;
  processes: SopProcess[];
  created_by: string;
  created_at: string;
}

export interface SopDocumentFormValues {
  title: string;
  category_id: string;
  raw_text: string;
}

File: backend/app/schemas/sop.py
import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class SopCategoryCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    description: str | None = None


class SopCategoryResponse(BaseModel):
    id: uuid.UUID
    slug: str
    name: str
    description: str | None = None

    model_config = {"from_attributes": True}


class SopProcessResponse(BaseModel):
    id: uuid.UUID
    process_key: str
    title: str
    description: str
    trigger_phrases: list[str]
    steps: list[dict[str, Any]]
    tools: list[str]
    sort_order: int

    model_config = {"from_attributes": True}


class SopDocumentCreate(BaseModel):
    title: str
    category_id: uuid.UUID
    raw_text: str


class SopDocumentUpdate(BaseModel):
    title: str
    category_id: uuid.UUID
    raw_text: str


class SopDocumentEnabledUpdate(BaseModel):
    enabled: bool


class SopDocumentListItem(BaseModel):
    id: uuid.UUID
    title: str
    category_id: uuid.UUID
    category_slug: str | None = None
    category_name: str | None = None
    summary: str | None
    enabled: bool
    processing_status: str
    processing_error: str | None
    process_count: int = 0
    tool_warnings: list[str] | None = None
    updated_at: datetime

    model_config = {"from_attributes": True}


class SopDocumentResponse(BaseModel):
    id: uuid.UUID
    title: str
    category_id: uuid.UUID
    category_slug: str | None = None
    category_name: str | None = None
    raw_text: str
    summary: str | None
    markdown: str | None
    content_hash: str
    enabled: bool
    processing_status: str
    processing_error: str | None
    tool_warnings: list[str] | None = None
    embedding_model: str | None
    process_count: int = 0
    processes: list[SopProcessResponse] = Field(default_factory=list)
    created_by: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}
